File: android/app/src/main/java/com/arbitrage/engine/domain/model/Contracts.kt
package com.arbitrage.engine.domain.model

/**
 * Shared domain models — the Kotlin mirror of backend/schemas.py and
 * backend/api/contracts.py. Keep field names/types in sync with those files.
 */

enum class TradingMode { TESTNET, LIVE }

enum class ExecutionStatus { SUCCESS, PARTIAL_FILL, HEDGED, REJECTED }

data class ArbitrageSignal(
    val executionProbability: Double,
    val expectedAlphaBps: Double,
    val adverseHazard: Double
)

data class RiskLimits(
    val minAlphaBps: Double = 15.0,
    val minExecutionProbability: Double = 0.85,
    val maxAdverseHazard: Double = 0.20,
    val maxNotionalUsdPerTrade: Double = 50.0,
    val maxDailyNotionalUsd: Double = 500.0,
    val killSwitchEngaged: Boolean = false
)

data class PortfolioBalance(
    val exchange: String,
    val asset: String,
    val free: Double,
    val locked: Double
)

data class TradeExecutionResult(
    val status: ExecutionStatus,
    val buyExchange: String,
    val sellExchange: String,
    val symbol: String,
    val executedVolumeUsd: Double,
    val grossSpreadPct: Double,
    val netSpreadPct: Double,
    val realizedPnlUsd: Double,
    val mlConfidenceScore: Double,
    val reason: String? = null
)

data class TradeSignalRequest(
    val userId: String,
    val symbol: String = "BTC/USDT",
    val exchangeBuy: String,
    val exchangeSell: String,
    val capitalAllocationUsd: Double,
    val minAlphaBps: Double = 15.0,
    val tradingMode: TradingMode = TradingMode.TESTNET
)

File: backend/api/contracts.py
"""Pydantic request/response contracts for the API gateway (unit 5).

Other units (marketdata, ml, execution, security) should not import this
module — it depends on them, never the other way around, to avoid an import
cycle when the API wires everything together.
"""
from __future__ import annotations

from pydantic import BaseModel, Field

from backend.schemas import ExecutionStatus, TradingMode


class TradeSignalRequest(BaseModel):
    user_id: str
    symbol: str = "BTC/USDT"
    exchange_buy: str
    exchange_sell: str
    capital_allocation_usd: float = Field(gt=0)
    min_alpha_bps: float = 15.0
    trading_mode: TradingMode = TradingMode.TESTNET


class TradeSignalResponse(BaseModel):
    status: str  # "ORDER_DISPATCHED" | "SIGNAL_REJECTED"
    reason: str | None = None
    metrics: dict
    allocated_capital: float | None = None
    target_pair: str | None = None


class ExecutionRecord(BaseModel):
    id: str
    user_id: str
    symbol: str
    buy_exchange: str
    sell_exchange: str
    gross_spread_pct: float
    net_spread_pct: float
    executed_volume_usd: float
    realized_pnl_usd: float
    ml_confidence_score: float
    execution_status: ExecutionStatus
